### PR TITLE
stage2: error set type equality, error and error union value equality

### PR DESCRIPTION
**Pull request by mitchellh**
This implements type equality for merged error types. This is done through element-wise error set comparison. This is a bit odd since no other error type is structurally compared (see #11022) but it appears to be the same behavior that stage1 also expects based on tests.

Also took the opportunity to clean up the style of error set comparisons.

This gets 6 more error-specific tests passing!

### CODE REVIEW

File: src/type.zig
Comment (andrewrk) [inline src/type.zig:570]: this looks wrong - why can't the tags be different?

**Review by andrewrk — CHANGES_REQUESTED** [review]

File: src/type.zig
Comment (mitchellh) [reply src/type.zig:570]: This is the same logic as before (see split diff!)

Comment (andrewrk) [reply src/type.zig:570]: the `return false` at the end of master branch is wrong - it's essentially returning a false negative as a TODO.

Comment (mitchellh) [reply src/type.zig:570]: I can turn this into `unreachable` for now if that'll unblock this PR, since it feels like merging the changes of #11022 into this might be heavy (I have those in a branch, but requires more changes i.e. due to resolving inferred error sets).

Comment (andrewrk) [reply src/type.zig:570]: Alright, #11022 is marked "accepted" now, hope that helps :-)

I'm fine with however you want to structure the changesets.

Also I think we can probably get away with implementing this proposal only in stage2 and leaving stage1 to rot for now. In case you haven't seen it, this is the plan to deal with that: #6378

Comment (mitchellh) [reply src/type.zig:570]: Hah okay, stand by. I'll re-request review once those are in this PR.

**Comment by mitchellh** [comment]
I added the logic for #11022. I'm not sure why the drone CI is failing on the `-ofmt=c` run.

File: src/type.zig
Comment (mitchellh) [reply src/type.zig]: Added it, note that even though the changes are only in stage2, I don't think anything was or is exercising the changed behavior in the compiler (besides new tests). So in regards to #6378 I don't think this will negatively impact bootstrapping in any way.

**Comment by andrewrk** [comment]
Likely the different value returned for equality caused some different codepath somewhere and uncovered a missing feature or bug in the C backend. Probably it's just a couple test cases that can now be regressed. I'd be happy to take it from here if you want.

**Comment by mitchellh** [comment]
Yes please all yours! since I don't have the local env setup (or knowledge yet) to debug the C backend. 😄

File: src/type.zig
Comment (andrewrk) [inline src/type.zig:570]: This can return a false negative:

```zig
    /// Returns true if it is an error set that includes anyerror, false otherwise.
    /// Note that the result may be a false negative if the type did not get error set
    /// resolution prior to this call.
    pub fn isAnyError(ty: Type) bool {
```

File: src/type.zig
Comment (andrewrk) [inline src/type.zig:582]: This could be a false negative.

**Review by andrewrk — CHANGES_REQUESTED** [review]
I think `Type.eql` should never return false negatives because that would cause subtle bugs for us down the road. There's also `Type.hash` to implement along with these changes.

File: src/type.zig
Comment (mitchellh) [reply src/type.zig:570]: Ah yes, I should do this later in the call.

**Comment by andrewrk** [comment]
I suggest to delete this from test/stage2/x86_64.zig:

```zig
        {
            var case = ctx.exe("error set equality", target);

            case.addCompareOutput(
                \\pub fn main() void {
                \\    assert(@TypeOf(error.Foo) == @TypeOf(error.Foo));
                \\    assert(@TypeOf(error.Bar) != @TypeOf(error.Foo));
                \\    assert(anyerror == anyerror);
                \\    assert(error{Foo} != error{Foo});
                \\    // TODO put inferred error sets here when @typeInfo works
                \\}
                \\fn assert(b: bool) void {
                \\    if (!b) unreachable;
                \\}
            ,
                "",
            );
        }
```

it doesn't belong there, and we already have coverage for this in behavior tests. these kind of tests are supposed to be for testing incremental compilation and other complex invocations, but the test harness got abused early on before we had `zig test` working.

File: src/Module.zig
Comment (andrewrk) [inline src/Module.zig:863]: I made #11109 so that this can be improved:

```suggestion
    pub fn sortNames(names: *NameMap) void {
        const Context = struct {
            keys: [][]const u8,

            pub fn lessThan(ctx: @This(), a_index: usize, b_index: usize) bool {
                return std.mem.lessThan(u8, ctx.keys[a_index], ctx.keys[b_index]);
            }
        };
        names.sort(Context{ .keys = names.keys() });
    }
```

**Review by andrewrk — CHANGES_REQUESTED** [review]

File: src/Module.zig
Comment (mitchellh) [reply src/Module.zig:863]: Looks great! Once #11109 merges I'll make those changes.

Comment (mitchellh) [reply src/Module.zig]: Done!

**Comment by mitchellh** [comment]
Using the new sort method, should be good to go.

**Pull request merged**
